server/src/main/java/io/crate/role/Roles.java

```diff
@@ -32,26 +32,31 @@
 
 public interface Roles {
 
-    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, Boolean> HAS_PRIVILEGE_FUNCTION = (r, t, c, o) ->
-        r.privileges().matchPrivilege(t, c, (String) o);
+    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, PrivilegeState> HAS_PRIVILEGE_FUNCTION =
+        (r, t, c, o) -> r.privileges().matchPrivilege(t, c, (String) o);
 
-    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, Boolean> HAS_ANY_PRIVILEGE_FUNCTION = (r, t, c, o) ->
-        r.privileges().matchPrivilegeOfAnyType(c, (String) o);
+    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, PrivilegeState> HAS_ANY_PRIVILEGE_FUNCTION =
+        (r, t, c, o) -> r.privileges().matchPrivilegeOfAnyType(c, (String) o);
 
-    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, Boolean> HAS_SCHEMA_PRIVILEGE_FUNCTION =
+    FourFunction<Role, Privilege.Type, Privilege.Clazz, Object, PrivilegeState> HAS_SCHEMA_PRIVILEGE_FUNCTION =
         (r, t, c, o) -> {
+            PrivilegeState result = PrivilegeState.REVOKE;
             for (Privilege privilege : r.privileges()) {
-                if (privilege.state() == PrivilegeState.GRANT && privilege.ident().type() == t) {
-                    if (privilege.ident().clazz() == Privilege.Clazz.CLUSTER) {
-                        return true;
-                    }
-                    if (privilege.ident().clazz() == Privilege.Clazz.SCHEMA &&
-                        OidHash.schemaOid(privilege.ident().ident()) == (Integer) o) {
-                        return true;
+                if (privilege.ident().type() == t) {
+                    if (privilege.ident().clazz() == Privilege.Clazz.SCHEMA
+                        && OidHash.schemaOid(privilege.ident().ident()) == (Integer) o) {
+
+                        if (privilege.state() == PrivilegeState.DENY || privilege.state() == PrivilegeState.GRANT) {
+                            return privilege.state();
+                        }
+                    } else if (privilege.ident().clazz() == Privilege.Clazz.CLUSTER) {
+                        if (result == PrivilegeState.REVOKE) {
```

Comment:
Similar here, is this needed? I think the result could always be overwritten, having multiple cluster privileges of the same type shouldn't be possible anyhow.

Reply:
Just writing down the thinking behind it to validate with you: the issue is that we are checking by schema oid (used in HasPrivilegeFunction), so we need to iterate over all privileges of the user/role (in arbitrary order of schemas and cluster) to find the schema matching. If we find it as you correctly spotted in your other comment we return. But if didn't find we need to know what is the state for `CLUSTER` clazz, and after the loop return it. And yes, indeed we don't need the if.

Reply:
> But if didn't find we need to know what is the state for CLUSTER clazz, and after the loop return it. And yes, indeed we don't need the if.

Yep, and also the other way around, a cluster privilege is found first, but a schema privilege is found afterwards. thus we cannot return the cluster state immediately but only if nothing else is found.